### ISSUE TITLE
syncoid progress bar is broken

### ISSUE THREAD

**Issue by devZer0** [jimsalterjrs/sanoid#122]
i would expect that it shows 100% for every zfs fs finished - but for most it doesn`t but shows arbitrary values after finishing transfer. sometimes much smaller (e.g. 38%) somtimes much larger (329%)

[root@backupvm1]# ./syncoid -r zfspool/backup root@targethost:zfspool/backup
Sending incremental zfspool/backup@syncoid_backupvm1_2017-08-03:10:22:20 ... syncoid_backupvm1._2017-08-04:09:59:39 (~ 4 KB):
1,52kiB 0:00:00 [3,23kiB/s] [==============================================================>                                                                                                       ] 38%
Sending incremental zfspool/backup/adminvm@syncoid_backupvm1_2017-08-03:10:22:27 ... syncoid_backupvm1_2017-08-04:09:59:46 (~ 17.7 MB):
18,1MiB 0:00:00 [25,8MiB/s] [=====================================================================================================================================================================] 102%
Sending incremental zfspool/backup/mysql@syncoid_backupvm1_2017-08-03:10:22:35 ... syncoid_backupvm1_2017-08-04:09:59:51 (~ 43.7 MB):
44,5MiB 0:00:01 [39,8MiB/s] [=====================================================================================================================================================================] 101%
Sending incremental zfspool/backup/mysql2@syncoid_backupvm1_2017-08-03:10:22:41 ... syncoid_backupvm1_2017-08-04:09:59:56 (~ 10.0 MB):
10,5MiB 0:00:00 [12,4MiB/s] [=====================================================================================================================================================================] 105%
Sending incremental zfspool/backup/server1@syncoid_backupvm1_2017-08-03:10:22:45 ... syncoid_backupvm1_2017-08-04:10:00:00 (~ 4 KB):
2,13kiB 0:00:00 [6,24kiB/s] [======================================================================================>                                                                              ] 53%
Sending incremental zfspool/backup/server2@syncoid_backupvm1_2017-08-03:10:22:48 ... syncoid_backupvm1_2017-08-04:10:00:02 (~ 101.5 MB):
 102MiB 0:00:02 [  42MiB/s] [====================================================================================================================================================================>] 100%
Sending incremental zfspool/backup/server3@syncoid_backupvm1_2017-08-03:10:22:54 ... syncoid_backupvm1_2017-08-04:10:00:08 (~ 5.8 MB):
 6,6MiB 0:00:00 [8,96MiB/s] [=====================================================================================================================================================================] 113%
Sending incremental zfspool/backup/server4@syncoid_backupvm1_2017-08-03:10:22:59 ... syncoid_backupvm1_2017-08-04:10:00:10 (~ 4 KB):
2,13kiB 0:00:00 [6,83kiB/s] [======================================================================================>                                                                              ] 53%
Sending incremental zfspool/backup/server4@syncoid_backupvm1_2017-08-03:10:23:03 ... syncoid_server5_2017-08-04:10:00:13 (~ 906.3 MB):
 975MiB 0:00:19 [49,2MiB/s] [=====================================================================================================================================================================] 107%
Sending incremental zfspool/backup/server6@syncoid_backupvm1_2017-08-03:10:24:36 ... syncoid_backupvm1_2017-08-04:10:00:38 (~ 3.2 GB):
 302MiB 0:00:05 [68,6MiB/s] [=============>                                                                                                                                                        ]  9% ETA 0:00:48

<snip>

Sending incremental zfspool/backup/server10@syncoid_backupvm1_2017-08-03:10:29:46 ... syncoid_backupvm1_2017-08-04:10:03:02 (~ 403.4 MB):
 404MiB 0:00:05 [74,2MiB/s] [====================================================================================================================================================================>] 100%
Sending incremental zfspool/backup/server11@syncoid_backupvm1_2017-08-03:10:29:58 ... syncoid_backupvm1_2017-08-04:10:03:12 (~ 172.5 MB):
 173MiB 0:00:02 [64,3MiB/s] [====================================================================================================================================================================>] 100%
Sending incremental zfspool/backup/server12@syncoid_backupvm1_2017-08-03:10:30:09 ... syncoid_backupvm1_2017-08-04:10:03:19 (~ 29.3 MB):
29,8MiB 0:00:01 [27,3MiB/s] [=====================================================================================================================================================================] 101%
Sending incremental zfspool/backup/server13@syncoid_backupvm1_2017-08-03:10:30:16 ... syncoid_backupvm1_2017-08-04:10:03:25 (~ 385.2 MB):
 385MiB 0:00:04 [80,7MiB/s] [====================================================================================================================================================================>] 100%
Sending incremental zfspool/backup/server14@syncoid_backupvm1_2017-08-03:10:30:25 ... syncoid_backupvm1_2017-08-04:10:03:34 (~ 1.8 MB):
1,91MiB 0:00:00 [ 2,6MiB/s] [=====================================================================================================================================================================] 103%
Sending incremental zfspool/backup/server15@syncoid_backupvm1_2017-08-03:15:13:27 ... syncoid_backupvm1_2017-08-04:10:03:44 (~ 9 KB):
31,3kiB 0:00:00 [65,7kiB/s] [=====================================================================================================================================================================] 329%
Sending incremental zfspool/backup/server16@syncoid_backupvm1_2017-08-03:15:13:54 ... syncoid_backupvm1_2017-08-04:10:03:49 (~ 3.8 MB):
4,15MiB 0:00:00 [7,27MiB/s] [=====================================================================================================================================================================] 110%
Sending incremental zfspool/backup/server17@syncoid_backupvm1_2017-08-03:15:15:06 ... syncoid_backupvm1_2017-08-04:10:03:54 (~ 11 KB):
40,9kiB 0:00:00 [90,2kiB/s] [=====================================================================================================================================================================] 371%
Sending incremental zfspool/backup/server18@syncoid_backupvm1_2017-08-03:15:15:33 ... syncoid_backupvm1_2017-08-04:10:03:59 (~ 42.4 MB):
  43MiB 0:00:00 [49,8MiB/s] [=====================================================================================================================================================================] 101%
Sending incremental zfspool/backup/server19@syncoid_backupvm1_2017-08-03:15:17:26 ... syncoid_backupvm1_2017-08-04:10:04:03 (~ 16.9 MB):
17,5MiB 0:00:00 [24,4MiB/s] [=====================================================================================================================================================================] 103%
Sending incremental zfspool/backup/server20@syncoid_backupvm1_2017-08-03:15:18:05 ... syncoid_backupvm1_2017-08-04:10:04:09 (~ 16.8 MB):
17,3MiB 0:00:00 [25,1MiB/s] [=====================================================================================================================================================================] 103%
Sending incremental zfspool/backup/server21@syncoid_backupvm1_2017-08-03:15:18:40 ... syncoid_backupvm1_2017-08-04:10:04:15 (~ 24.6 MB):
25,6MiB 0:00:00 [32,2MiB/s] [=====================================================================================================================================================================] 103%
Sending incremental zfspool/backup/server22@syncoid_backupvm1_2017-08-03:15:19:42 ... syncoid_backupvm1_2017-08-04:10:04:20 (~ 31.0 MB):
32,3MiB 0:00:00 [34,6MiB/s] [=====================================================================================================================================================================] 104%
Sending incremental zfspool/backup/server23@syncoid_backupvm1_2017-08-03:15:20:29 ... syncoid_backupvm1_2017-08-04:10:04:26 (~ 32.8 MB):
  36MiB 0:00:01 [32,8MiB/s] [=====================================================================================================================================================================] 109%
Sending incremental zfspool/backup/server24@syncoid_backupvm1_2017-08-03:15:24:24 ... syncoid_backupvm1_2017-08-04:10:04:35 (~ 7.5 GB):
7,53GiB 0:02:02 [62,8MiB/s] [====================================================================================================================================================================>] 100%
Sending incremental zfspool/backup/server25@syncoid_backupvm1_2017-08-03:16:11:38 ... syncoid_backupvm1_2017-08-04:10:06:45 (~ 4 KB):
2,13kiB 0:00:00 [6,29kiB/s] [======================================================================================>                                                                               ] 53%
Sending incremental zfspool/backup/server26@syncoid_backupvm1_2017-08-03:16:12:28 ... syncoid_backupvm1_2017-08-04:10:06:50 (~ 14.2 MB):
  15MiB 0:00:00 [21,7MiB/s] [=====================================================================================================================================================================] 105%

**Comment by jimsalterjrs** [comment]
You're going to need to look upstream (to zfsonlinux at the least, but more likely all the way up to openzfs itself) for a fix. Syncoid uses the dry-run argument to zfs send first for zfs to tell it how much data a snapshot stream will contain, and uses that information to gauge its progress when sending the real stream.

As you're seeing here, with very small sync amounts (less than 100MB or so) you're likely to be off a few percent between what dry-run estimates, and the size of the actual stream when sent. Your choices are to live with that as a minor annoyance, or to file a bug with the zfs developers about it and see whether they can or will increase the estimation efficiency for very small stream sizes.

I'm going to go ahead and close this here, as it's not something I can do anything about - like I said, upstream bug.

**Comment by kpande** [comment]
no, this is not an upstream issue since your application is monitoring send progress incorrectly. use `send -p` to get machine-parseable output and write it to a file to detect progress from..

**Comment by devZer0** [comment]
mhh - this example shows "size 6753805752" and then it counts 6749076424 -> 6828295304 -> 6961438080, which is larger value than initial size reported.

for me it looks that the size estimation is wrong, too.

the question is if this will also make a progress bar calculated from that values to go nuts like i observed for sandoid

 [root@rx300-1 man8]# zfs send -vP zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300 |zfs recv -F zfspool/test
full	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300	6753805752
size	6753805752
17:35:06	1001832	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:07	151254256	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:08	338712728	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:09	514963200	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:10	772540832	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:11	1072302128	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:12	1219514416	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:13	1364762568	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:14	1374879136	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:15	1407945272	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:16	1461085888	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:17	1611911568	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:18	1858586664	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:19	2199789656	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:20	2356374712	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:21	2357350160	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:22	2413539944	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:23	2535287912	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:24	2778208576	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:25	2836539112	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:26	2922118536	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:27	3007724224	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:28	3230676768	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:29	3247694040	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:30	3257609936	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:31	3375385680	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:32	3466278072	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:33	3533177752	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:34	3603006912	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:35	3744648120	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:36	3805736552	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:37	3872383672	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:38	3946167472	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:39	4028965160	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:40	4125227856	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:41	4222307616	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:42	4257978712	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:43	4305984272	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:44	4393539416	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:45	4483999760	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:46	4525236928	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:47	4598830840	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:48	5160172960	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:49	5299995672	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:50	5352386624	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:51	5559490840	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:52	5733608632	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:53	5750637736	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:54	5802192184	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:55	5873150880	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:56	6129140992	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:57	6381630736	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:58	6447362864	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:35:59	6749076424	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:36:00	6828295304	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300
17:36:01	6961438080	zfspool/backup/server1@zfs-auto-snap_daily-2017-07-20-2300

**Comment by jimsalterjrs** [comment]
Yes, there is a simplistic fudge factor in sanoid that attempts to compensate for dry run's typically low estimates. This is very obvious in the code. Again, I can't fix this, you need to take it up with OpenZFS upstream.

**Comment by jimsalterjrs** [comment]
@kpande did you actually look at the code? It's not particularly hard to follow.

````
        my $getsendsizecmd = "$sourcessh $mysudocmd $zfscmd send -nP $snaps";
        if ($debug) { print "DEBUG: getting estimated transfer size from source $sourcehost using \"$getsendsizecmd 2>&1 |\"...\n"; }

        open FH, "$getsendsizecmd 2>&1 |";
        my @rawsize = <FH>;
        close FH;
        my $exit = $?;

        # process sendsize: last line of multi-line output is 
        # size of proposed xfer in bytes, but we need to remove 
        # human-readable crap from it 
        my $sendsize = pop(@rawsize);
        $sendsize =~ s/^size\s*//;
        chomp $sendsize;

        # to avoid confusion with a zero size pv, give sendsize
        # a minimum 4K value - or if empty, make sure it reads UNKNOWN
        if ($debug) { print "DEBUG: sendsize = $sendsize\n"; }
        if ($sendsize eq '' || $exit != 0) {
                $sendsize = '0';
        } elsif ($sendsize < 4096) {
                $sendsize = 4096;
        }
        return $sendsize;
````

This is how we determine the send size to feed to the progress bar - we pull `zfs send -nP` against the same operation we'll do USING the progress bar; then if the estimate is below 4K, we set it to 4K (since `pv` really doesn't like getting called with a 0 `size` value).

This gets fed into `pv` directly inside `buildsynccmd()`:

~~~~
if ($avail{'localpv'} && !$quiet) { $synccmd .= "$pvcmd -s $pvsize | "; }
~~~

Again: **there is nothing I can do about this**. If you want a more accurate progress bar, you'll need more accurate dry-run estimates, and that needs to be taken up with zfsonlinux, or more likely openzfs. I do not develop the zfs codebase itself and cannot help directly with that.

**Comment by jimsalterjrs** [comment]
@kpande If you want a very direct example, here you go:

````
root@banshee:/banshee/target# zfs send -nP -I data/smbshare@autosnap_2017-07-08_23:59:01_daily data/smbshare@autosnap_2017-08-03_23:59:01_daily | tail -n 1
size	925864
root@banshee:/banshee/target# zfs send -I data/smbshare@autosnap_2017-07-08_23:59:01_daily data/smbshare@autosnap_2017-08-03_23:59:01_daily | wc -c
744600
````
`zfs send -nP` reports 925,864 bytes; actual `zfs send` produces 744,600 bytes of data. This operation would result in an inaccurate progress bar in `syncoid`, just as it's inaccurate here on the command line.

**Comment by kpande** [comment]
you're doing it wrong, `send -P` actually output the current progress point. this gives me 95% of the total sent instead of 80% using your method.

**Comment by kpande** [comment]
the 5% missing is because the output only prints every second and the send completed before the final output. if anything, that is the upstream "bug".

**Comment by jimsalterjrs** [comment]
@kpande, that's a send of a single snapshot. Please try again with an incremental stream, as demonstrated. For a *large* delta, the `-nP` estimate is reasonably accurate - enough to get a 100% out of a `pv` progress bar, at least.  For small deltas, it can be inaccurate by as much as 300%.  This is where the inaccuracy in `syncoid`'s progress bar comes from - inaccurate size data returned from `zfs send -nP -I`.

Incidentally, the `-c` flag for send is not yet supported in the ZFS versions available in production. Current in Ubuntu is 0.6.5.6-oubuntu17. I don't believe 0.7.0 is production in BSD yet either. I'll certainly start using it once 0.7.0 is mainstream, but we're not there yet.

**Comment by kpande** [comment]
zfsonlinux 0.7.0 is released.

**Comment by jimsalterjrs** [comment]
I know it's "released" as in "available for download and compile from source yourself", but that's not "released" as in "in production repos".

Please demonstrate your -P again with the *actual* parameters used - `zfs send -nP -I snap1 snap2`. I cannot just pipe potentially 20T of data through to /dev/null in order to get a total size so I can give a progress bar with percentage completion.

If -nPI works correctly in 0.7.0, then Sanoid's progress bar will also work correctly with 0.7.0.  Right now, on 0.6.5.6, it does not work correctly in all cases, because -nPI does not work correctly in all cases.

**Comment by kpande** [comment]
you don't have to pipe to /dev/null, it prints the size just before it starts sending and you can read this arbitrarily if you write to a file (`2> /tmp/file`): it says "size ...." as the 2nd line printed, and then the `-v` argument to `send` will have it print a status line every 1 second. there is an issue where <1 second interval of data is not counted at the very end.


this can be fixed upstream but it's not as big of a deal as compared to the more fundamental problem, how you are counting data via `pv` instead of the /actual/ total from `zfs send -P`. in fact, using pv is a bad idea because of unnecessary memcpy overhead. if I were to use anything it would be mbuffer to provide a stable pool for send/recv to avoid getting hung up by linux's relatively tiny pipe buffer. but even this is unnecessary.

even if I add the `-n` to `send`, I get the same result, so dry run isn't at fault here either:

```
# zfs send -nvP -I tank/obj@send tank/obj@send2                       
incremental     send    tank/obj@send2  624
size    624
```

**Comment by jimsalterjrs** [comment]
pv is capable of moving data at gigabytes per second. Not a bottleneck.

The real question here is why -nP produces incorrect output, if -nvP produces correct output.

**Comment by kpande** [comment]
it is using extra CPU which limits the throughput in very fast environments, there is no question about this.

both produce the same output, the only difference is that `n` doesn't produce progress text since there is no actual send going.

**Comment by kpande** [comment]
are you using dedup?

**Comment by jimsalterjrs** [comment]
I am not using dedup, and I'm open to suggestions as to why it's *not* producing the correct output. Again, it's possible that this very long standing bug in -nP has been corrected since 0.6.5.6 and 0.7.0, but you can see for yourself that it does not, in fact, produce the correct size output in 0.6.5.6.

This behavior has been consistent for around five years at least. On a large send, it's close enough to accurate - within half a percent, on sends of several hundred MB. For small sends - again - it's pretty inaccurate, and has been for several years, on the >100 servers I manage in production, all of which use replication on a daily (or more frequent) basis.

**Comment by kpande** [comment]
but until you are using the `send -Pv` output to track instead of `pv` then you can not say it is a bug in ZFS upstream.

**Comment by jimsalterjrs** [comment]
The dataset you can see -nP producing an incorrect estimate of stream size above has no dedup, lz4 compression, xattr=sa. On mobile right now so can't remember for sure if recordsize is 8K or the default 128K on that particular dataset, but I can tell you the behavior has been consistent on datasets of both recordsizes. It's also been consistent on ashift values of 9, 12, and 13 - that particular vdev is ashift=9.

**Comment by kpande** [comment]
is it possible that you are using zvol?

https://github.com/zfsonlinux/zfs/pull/6205

**Comment by jimsalterjrs** [comment]
You are completely incorrect. I demonstrated the incorrect size above, no use of pv required. Unless you'd like to suggest that wc and pv share the same bug?

**Comment by jimsalterjrs** [comment]
I am not using zvols.

**Comment by heeplr** [comment]
> Again: there is nothing I can do about this. If you want a more accurate progress bar, you'll need more accurate dry-run estimates, and that needs to be taken up with zfsonlinux, or more likely openzfs. I do not develop the zfs codebase itself and cannot help directly with that.

I encountered this issue and found that ZFS' size guesstimate is completely off mainly on small sends which wouldn't need a progress bar anyway.

@jimsalterjrs Maybe it would be possible to not show a progress bar (omit --size argument) for sizes below a certain threshold. That would eliminate the confusing status output without degrading the user experience.
